### PR TITLE
Remove alert system

### PR DESCRIPTION
**Pull request by laanwj**
This completely removes the P2P network alert system.

Internal "alerts", such as fork warnings, along with the `-alertnotify` option are kept.

Removes the last vestiges of centralization, and puts an end to discussions such as #5160.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -1226,6 +1225,26 @@ bool fLargeWorkForkFound = false;
...
+    std::string strCmd = GetArg("-alertnotify", "");
+    if (strCmd.empty()) return;
+
+    // Alert text should be plain ascii coming from a trusted source, but to
+    // be safe we first strip anything not in safeChars, then add single quotes around
+    // the whole string before passing it to the shell:
+    std::string singleQuote("'");
+    std::string safeStatus = SanitizeString(strMessage);
+    safeStatus = singleQuote+safeStatus+singleQuote;
+    boost::replace_all(strCmd, "%s", safeStatus);
```
Comment (petertodd) [inline src/main.cpp]: Do we really need to keep the alert text sanitization?

Comment (laanwj) [inline src/main.cpp]: I think so. It's just belt and suspenders, prudent in case a) some external string leaks into the warning b) some message from inside the program accidentally results in a shell escape.

Comment (petertodd) [inline src/main.cpp]: Sounds like a reasonable explanation to me.

**Comment by petertodd** [comment]
utACK

**Comment by jgarzik** [comment]
hmmmm.  I agree w/ the theory.

To be practical, sites either implicitly or explicitly rely on this to save their pants in some cases, like it or not.

Yanking it without replacement or user communication negatively impacts current users and leaves them less protected versus today.

To proceed, something like an optional PGP-based alert network should be in place, with an initial membership list similar to bitcoin-security.  With that in place, I could ACK this change.

**Comment by petertodd** [comment]
@jgarzik I suggested awhile back that we use transactions themselves as the broadcast medium for alerts; authenticate via scriptPubKey and put the message in an OP_RETURN. If miners are censoring the txs, not a big deal, if they aren't you have strong guarantees that you haven't missed any. Finally, spam is handled by fees, so you don't need need whitelists of who's allowed to make an alert.

**Comment by laanwj** [comment]
@jgarzik An alternative notification system would be useful; as you say, it could just be a mailing list people can subscribe to. The transaction-based system could be useful too, although it would need to be external to the core client. No hardcoded whitelist anymore.

There is a pressing reasons why this centralized broadcast system should be removed from the core network and core client, though. There is no longer consensus (neither within nor outside) the development team as to what is the way forward for Bitcoin, making it a liability.

File: src/test/alert_tests.cpp [hunk 2]
```diff
@@ -1,259 +0,0 @@
...
-    BOOST_CHECK_EQUAL(r[3], "'Evil Alert; /bin/ls; echo ' ");
-#endif
-    boost::filesystem::remove(temp);
-
-    SetMockTime(0);
-}
-
-static bool falseFunc() { return false; }
-
-BOOST_AUTO_TEST_CASE(PartitionAlert)
```
Comment (laanwj) [inline src/test/alert_tests.cpp]: This test is unrelated to the P2P alert system and should be kept

**Comment by jgarzik** [comment]
Either way it is only right to assess the impact from the point of view of current users and current field needs and expectations, regardless of where a moving-forward consensus may lie...

**Comment by wtogami** [comment]
As much as I agree with the good reasons to remove the alert system, I wonder if we should discuss alternatives first.

theymos mentioned major drawbacks that we would have if the alert system were to be completely removed.  For example, alerts have been (historically useful in times of emergency)[https://en.bitcoin.it/wiki/Alerts] to warn users about bugs.  Would we be able to quickly inform everyone quickly enough in the next emergency?

I am curious if we would be better off with some kind of improved alert system like n-of-m multisig.  For example, the (development demo Alpha sidechain)[https://github.com/ElementsProject/elementsproject.github.io/] currently is deployed with a simple 2-of-3 multisig alert system.  Multisig alerts would be an improvement over the current alert system in at least two ways:

* No single signer can abuse the alert system.
* Signer identities would be known to the community, so it could improve accountability.

Yes, it is a good question of who does the community trust with the signing keys, and what n-of-m multisig threshold is needed.  I am only saying maybe we should think about this harder before simple removal?

**Comment by petertodd** [comment]
The problem isn't having *a* alert system; the problem is we don't have a good way of having more than one alert system. n-of-m multisig helps this, but it's not a huge improvement.

Anyway, I don't see any technical downsides to using valid transactions as alerts; I'd be happy to code this up.

**Comment by wtogami** [comment]
It could be valid n-of-m multisig transactions that become alert carriers?

(I don't know if this is a good idea, just mentioning it.)

**Comment by petertodd** [comment]
@wtogami Yes exactly. Basically that provides a flood-fill mechanism with anti-spam protections, removing the need to have a whitelist of allowed alert signers. You'd just specify a P2SH address allowed to send alerts and put the alert itself in an 80-byte OP_RETURN.

Actually authenticating the alerts is easy for full nodes. For P2SH nodes, a neat trick is to take advantage of the fact that the P2SH redeemScript is in the scriptSig, which lets them both match it via bloom filters, and then re-run the script validation to validate the alert.

**Comment by gmaxwell** [comment]
I've been suggesting this be removed periodically for some time now.  It's a point of centralization we don't need, and don't benefit from-- one which creates other confusion and discord.  Multisig is nice but doesn't really address the fundamental problems with it.

If we wanted to have something to alert people that they were potentially on the "wrong" fork, something that caused a message upon burning 500 fork-BTC would be a lot more equitable.

**Comment by petertodd** [comment]
> If we wanted to have something to alert people that they were potentially on the "wrong" fork, something that caused a message upon burning 500 fork-BTC would be a lot more equitable.

That's a brilliant idea.

Though to do it well we really need a OP_GETBLOCKHASHVERIFY all all the fuss that entails. (possible too with spending recently mined coins, but kinda a pain to arrange)

**Comment by gmaxwell** [comment]
well you just move them first on the other chain.  Best to do when you're _really_ sure that that other chain isn't going to lose. :)

**Comment by petertodd** [comment]
@gmaxwell Indeed! Beng able to guarantee you won't lose unless it wins it much easier. :)

**Comment by theymos** [comment]
IMO alerts should still exist in some form for serious software bugs. For example, it seemed possible that heartbleed could have been used to steal bitcoins from users of Bitcoin-Qt in some cases (though apparently this was later learned to be unlikely). Users should be warned about this kind of thing, especially on Windows where there is no package manager and users expect their programs to auto-update themselves.

If there's any concern about alerts being misused, they could be restricted to a limited set of codes that the client expands into full warning messages (which can also be localized). For example, an alert with text "crit_vuln" might expand to "URGENT: This version may contain a critical security vulnerability. Shut it down right away and watch your favorite Bitcoin sites for news about updates. Be sure to verify the signatures of any new version before you run it."

**Comment by laanwj** [comment]
OK, closing this, too controversial.

**Comment by jtimon** [comment]
> If there's any concern about alerts being misused, they could be restricted to a limited set of codes that the client expands into full warning messages (which can also be localized). For example, an alert with text "crit_vuln" might expand to "URGENT: This version may contain a critical security vulnerability. Shut it down right away and watch your favorite Bitcoin sites for news about updates. Be sure to verify the signatures of any new version before you run it."

I think this (limit the alert messages that can be send) combined with bitcoin scripting for the signatures (to get multisig) would be a minimum to move to in case we wan't to maintain a centralized alert system for non-fork alerts.
Implementing @gmaxwell 's mechanism for fork alerts in parallel would be interesting too, of course.
In fact all 3 things can be implemented in parallel if there's interest on them. The later seems a minimum requirement for completely eliminating the centralized alerts.

**Comment by laanwj** [comment]
Continued in #6274.
